calyx-opt/src/passes/compile_static.rs

```diff
@@ -75,99 +87,265 @@ fn make_guard_dyn(
           builder;
+          // increments the fsm
           adder["left"] = ? fsm["out"];
           adder["right"] = ? const_one["out"];
-          fsm["write_en"] = not_last_state_guard ? signal_on["out"];
-          fsm["in"] = not_last_state_guard ? adder["out"];
-          // need done condition because we are creating a dynamic group
-          g["done"] = last_state_guard ? signal_on["out"];
+          fsm["write_en"] = ? signal_on["out"];
+          fsm["in"] = not_penultimate_state_guard ? adder["out"];
+           // resets the fsm early
+          fsm["in"] = penultimate_state_guard ? first_state["out"];
+          // will never reach this guard since we are resetting when we get to
+          // the penultimate state
+          g["done"] = ? ud["out"];
         );
         assigns.extend(fsm_incr_assigns.to_vec());
+        // maps the "early reset" group name to the "fsm name" that it borrows.
+        // this is helpful when we build the "wrapper group"
+        self.fsm_map.insert(g.borrow().name(), fsm.borrow().name());
         // adding the assignments to the new dynamic group and creating a
         // new (dynamic) enable
         g.borrow_mut().assignments = assigns;
-        g.borrow_mut().attributes = sgroup.attributes.clone();
-        let mut e = ir::Control::enable(g);
-        let attrs = std::mem::take(&mut s.attributes);
-        *e.get_mut_attributes() = attrs;
-        // need to add a continuous assignment to reset the fsm
-        let fsm_reset_assigns = build_assignments!(builder;
-            fsm["in"] = last_state_guard ? first_state["out"];
-            fsm["write_en"] = last_state_guard ? signal_on["out"];
+        g.borrow_mut().attributes = attributes;
+        g
+    }
+
+    fn build_wrapper_group(
+        fsm_name: &ir::Id,
+        group_name: &ir::Id,
+        builder: &mut ir::Builder,
+    ) -> ir::RRC<ir::Group> {
+        // get the groups/fsm necessary to build the wrapper group
+        let early_reset_group = builder
+            .component
+            .get_groups()
+            .find(*group_name)
+            .unwrap_or_else(|| {
+                panic!(
+                    "called build_wrapper_group with {}, which is not a group",
+                    group_name
+                )
+            });
+        let early_reset_fsm =
+            builder.component.find_cell(*fsm_name).unwrap_or_else(|| {
+                panic!(
+                    "called build_wrapper_group with {}, which is not an fsm",
+                    fsm_name
+                )
+            });
+        // not sure if there is a better way of getting the fsm width...
+        let fsm_width = early_reset_fsm
+            .borrow()
+            .ports()
+            .iter()
+            .find(|port| port.borrow().name == "in")
+            .unwrap_or_else(|| panic!("called {} in build_wrapper_group as an fsm; should have `in` port", early_reset_fsm.borrow().name()))
+            .borrow()
+            .width;
```

Comment:
This is probably not the best way to get the width of the FSM. An easier solution would be to pass the width to the function. Alternatively, you can look at the parameters of the parent cell of the FSM instead of looking at the width of the `in` port.